Typings `pyi` files added

GoodWasHere opened the pull request.

lelit commented on MANIFEST.in:
```diff
@@ -2,6 +2,7 @@ include .dir-locals.el
 include LICENSE MANIFEST.in *.cfg *.ini *.py *.rst *.txt
 include python-rapidjson/*.cpp python-rapidjson/*.h
 include benchmarks/*.py tests/*.py
+include typings/*.pyi
```
Why is this needed, given that AFAICT there's no `.pyi` there?

GoodWasHere replied:
My bad.
File extension fixed.

lelit commented:
Hi, sorry for the delay.
As I'm not an expert on these things, I wonder if there is a way to test this, in particular to avoid future divergences between the `.pyi` declarations and the actual implementation. What do you do/use to verify the declarations?

Spitfire1900 commented:
`stubtest` is suitable for this, but I am unsure how to best run it. 

This approach works and throws some lint warnings on definitions not available at runtime in the type file:

```shell
git clone --recursive https://github.com/python-rapidjson/python-rapidjson.git
cd python-rapidjson
mkdir typings
curl https://raw.githubusercontent.com/python-rapidjson/python-rapidjson/f0989a2b59a570368276adf21ebdd45df38ccac8/typings/__init__.pyi -o typings/rapidjson.pyi
curl https://raw.githubusercontent.com/python-rapidjson/python-rapidjson/f0989a2b59a570368276adf21ebdd45df38ccac8/MANIFEST.in -o MANIFEST.in
python -m venv .venv
source .venv/bin/activate
pip install .
cd typings
pip install mypy
stubtest rapidjson
```

outputs:
```
error: rapidjson.IM_ONLY_LISTS_TYPE is not present at runtime
Stub: in file C:\TMP\python-rapidjson\typings\rapidjson.pyi:24
Type alias for Literal[1]
Runtime:
MISSING

error: rapidjson.IterableMode is not present at runtime
Stub: in file C:\TMP\python-rapidjson\typings\rapidjson.pyi:89
Type alias for Union[Literal[0], Literal[1]]
Runtime:
MISSING

error: rapidjson.__rapidjson_version__ is not present in stub
Stub: in file C:\TMP\python-rapidjson\typings\rapidjson.pyi
MISSING
Runtime:
'1.1.0'

Found 49 errors (checked 1 module)
Return 1
```

Spitfire1900 commented:
I tried out stubgen on this project but it could not infer literal int contants, only that they were some int.

lelit commented:
Thank you for mentioning `stubgen`: from above I infer it comes with `mypy`...
I will try, as time permits.

GoodWasHere commented:
I apologize for the delay in replying: life attacked me.
I fixed the stubtest warnings.
But there is a problem: stubtest claims that in the runtime there are no next methods
```
class Decoder:
    def end_array(self, sequence: t.Sequence) -> t.Sequence: ...
    def end_object(self, mapping: t.Mapping) -> t.Mapping: ...
    def start_object(self) -> t.Union[t.Sequence, t.Mapping]: ...
    def string(self, s: str) -> str: ...

class Encoder:
    def default(self, obj: t.Any) -> JSONType: ...
```
I removed them from pyi for the sake of green status, but I don't understand what the reason is.

Spitfire1900 commented:
@GoodWasHere , how are you running stubtest BTW?

```
python-rapid-json @ stubtest rapidjson
error: not checking stubs due to failed mypy compile:
rapidjson/thirdparty/gtest/googlemock/scripts/fuse_gmock_files.py:235: error: Missing parentheses in call to 'print'. Did you mean print(...)?  [syntax]
Return 1
```

Spitfire1900 commented:
@GoodWasHere 

Please apply the following patch so that the "magic" `rapidjson-stubs` directory is created. PyLance and MyPy look for directories with this suffix to look for typing information. (I don't know where this is documented, this work was based on types-requests as a reference)

0001-include-stubs-in-rapidjson-stubs-directory.patch
```patch
From 8ed06ab2120e2570531543e7051da60d4f97d5ac Mon Sep 17 00:00:00 2001
From: Kyle Gottfried <6462596+Spitfire1900@users.noreply.github.com>
Date: Thu, 13 Jun 2024 16:46:59 -0400
Subject: [PATCH] include stubs in rapidjson-stubs directory

---
 setup.py | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/setup.py b/setup.py
index e86b1b2..d5f624a 100644
--- a/setup.py
+++ b/setup.py
@@ -115,5 +115,11 @@ setup(
         'Programming Language :: Python',
     ],
     ext_modules=[Extension('rapidjson', **extension_options)],
+    package_dir = {
+        'rapidjson-stubs': 'typings/rapidjson'
+    },
+    packages = ['rapidjson-stubs'],
+    package_data = {'rapidjson-stubs': ['*.pyi']},
+    include_package_data=True,
     **other_setup_options
 )
-- 
2.34.1
```

GoodWasHere commented:
@Spitfire1900 Done.

Spitfire1900 commented:
Found the documentation reference for `*-stubs` directories - https://peps.python.org/pep-0561/#stub-only-packages

Spitfire1900 commented:
I've also tested pyright's stub generator, it wasn't any better than stubgen.

Spitfire1900 commented:
@GoodWasHere , apply the following to ensure typing stubs remain up-to-date:

```diff
0001-add-typing-stub-tests.patch
From 0014e3123126e2fa6c0f29b75b515bcbf5c7b769 Mon Sep 17 00:00:00 2001
From: Kyle Gottfried <6462596+Spitfire1900@users.noreply.github.com>
Date: Fri, 14 Jun 2024 16:15:06 -0400
Subject: [PATCH] add typing stub tests

---
 .github/workflows/main.yml | 4 ++++
 requirements-test.txt      | 1 +
 2 files changed, 5 insertions(+)

diff --git a/.github/workflows/main.yml b/.github/workflows/main.yml
index 681fe96..1380436 100644
--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -52,6 +52,10 @@ jobs:
         run: |
           make -C docs doctest -e PYTHON=$(which python3)
 
+      - name: Typing stub tests
+        run: |
+          stubtest rapidjson
+
   debug-tests:
     name: Memory leak tests, on current debug build Python
     runs-on: ubuntu-latest
diff --git a/requirements-test.txt b/requirements-test.txt
index e454852..473c6e0 100644
--- a/requirements-test.txt
+++ b/requirements-test.txt
@@ -7,6 +7,7 @@
 # :Copyright: © 2015, 2017, 2018, 2021, 2022, 2023 Lele Gaifax
 #
 
+mypy==1.10.0
 pytest==7.4.3
 pytz==2023.3.post1
 sphinx==7.2.6
-- 
2.34.1
```

GoodWasHere commented:
Done.
But I'll remind that the following methods are removed from `.pyi`, since `stubtest` doesn't see them in reflection for some reason.
```
class Decoder:
    def end_array(self, sequence: t.Sequence) -> t.Sequence: ...
    def end_object(self, mapping: t.Mapping) -> t.Mapping: ...
    def start_object(self) -> t.Union[t.Sequence, t.Mapping]: ...
    def string(self, s: str) -> str: ...

class Encoder:
    def default(self, obj: t.Any) -> JSONType: ...
```

Spitfire1900 commented:
@GoodWasHere , I may be incorrect but that is because genuinely no bindings exist for those methods.

```python
>>> import rapidjson
>>> dir(rapidjson)
['BM_NONE', 'BM_UTF8', 'DM_IGNORE_TZ', 'DM_ISO8601', 'DM_NAIVE_IS_UTC', 'DM_NONE', 'DM_ONLY_SECONDS', 'DM_SHIFT_TO_UTC', 'DM_UNIX_TIME', 'Decoder', 'Encoder', 'IM_ANY_ITERABLE', 'IM_ONLY_LISTS', 'JSONDecodeError', 'MM_ANY_MAPPING', 'MM_COERCE_KEYS_TO_STRINGS', 'MM_ONLY_DICTS', 'MM_SKIP_NON_STRING_KEYS', 'MM_SORT_KEYS', 'NM_DECIMAL', 'NM_NAN', 'NM_NATIVE', 'NM_NONE', 'PM_COMMENTS', 'PM_NONE', 'PM_TRAILING_COMMAS', 'RawJSON', 'UM_CANONICAL', 'UM_HEX', 'UM_NONE', 'ValidationError', 'Validator', 'WM_COMPACT', 'WM_PRETTY', 'WM_SINGLE_LINE_ARRAY', '__author__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__rapidjson_version__', '__spec__', '__version__', 'dump', 'dumps', 'load', 'loads']
>>> rapidjson.Decoder()
<rapidjson.Decoder object at 0x7f3dd18991d0>
>>> d = rapidjson.Decoder()
>>> dir(d)
['__call__', '__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'datetime_mode', 'number_mode', 'parse_mode', 'uuid_mode']
>>> d.end_array
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'rapidjson.Decoder' object has no attribute 'end_array'
>>> 
```

GoodWasHere commented:
I get it, these methods can be defined in inheritors.
The signatures are taken from the [documentation](https://python-rapidjson.readthedocs.io/en/latest/decoder.html#rapidjson.Decoder.end_array).
I added exceptions for them, `stubtest` takes them into account when running
`python3.11 -m mypy.stubtest rapidjson --allowlist stubtest_allowlist.txt`

Spitfire1900 commented:
>I get it, these methods can be defined in inheritors.
The signatures are taken from the [documentation](https://python-rapidjson.readthedocs.io/en/latest/decoder.html#rapidjson.Decoder.end_array).
I added exceptions for them, stubtest takes them into account when running


My two cents. Since these methods are to be defined in inheritors and are not part of the base implementation they should not exist in the stubs. Instead it should come from the sub-type.


![image](https://github.com/python-rapidjson/python-rapidjson/assets/6462596/f0ece8a8-24bb-4254-8705-13129cf17523)
![image](https://github.com/python-rapidjson/python-rapidjson/assets/6462596/c2980d50-ab13-43e8-8b22-7ebb3ef4a690)

GoodWasHere commented:
It seems to me that the invisibility of a method in the parent is purely implementation peculiarities.
This is essentially an empty method in the parent that can be overridden.
And I think we should fix its signature so that mypy can check the signature of its descendants.

lelit commented:
> It seems to me that the invisibility of a method in the parent is purely implementation peculiarities. This is essentially an empty method in the parent that can be overridden.

With that, do you mean that the concrete implementation should change? I mean, [currently](https://github.com/python-rapidjson/python-rapidjson/blob/master/rapidjson.cpp#L832-L840) it checks for the presence of those methods, so there is not an _abstract_ implementantion...

lelit commented:
Provided we find a way to not introduce any overhead, inevitable when unconditionally calling a method... I've done that that way for a reason :wink:

lelit commented:
I will try to come up with either a measure of the overhead, or with a possible different approach. Not sure when, though, sorry.

lelit commented:
Ok, I quickly hacked one further _contender_ in the benchmarks, basically:
```python
class DecoderPython(rj.Decoder):
    def end_array(self, a):
        return a

    def start_object(self):
        return {}

    def end_object(self, d):
        return d

    def string(self, s):
        return s


class EncoderPython(rj.Encoder):
    def default(self, obj):
        return repr(obj)


contenders.append(Contender('rapidjson_p',
                            EncoderPython(),
                            DecoderPython()))
```

to obtain a comparison of that _empty method call_ overhead.

In the following tables, focus on the difference between, say `rapidjson_c-100 arrays dict` and `rapidjson_p-100 arrays dict`: the former executes the test against the native **C** encoder/decoder, while the latter against a _do-nothing-else_ **Python** subclass.

I removed almost all _deserialization_ tests, leaving just a few because the impact there is negligible.

As you can see the impact, when the test exercises those methods, **is** _appreciable_, IMHO.

### benchmark '100 arrays dict: deserialize': 5 tests

|Name (time in ms)                            |       Min          |       Max          |      Mean          |   StdDev          |    Median         |       IQR         |   Outliers |    OPS         |   Rounds | Iterations|
|---------------------------------------------|--------------------|--------------------|--------------------|-------------------|-------------------|-------------------|------------|----------------|----------|-----------|
|test_loads[rapidjson_nn_c-100 arrays dict]   |  415.5141 (1.0)    |  468.1543 (1.00)   |  440.0577 (1.00)   |  25.2165 (1.38)   |  434.2982 (1.0)   |   48.5943 (1.46)  |        1;0 | 2.2724 (1.00)  |        5 |          1|
|test_loads[rapidjson_nn_f-100 arrays dict]   |  417.4602 (1.00)   |  466.2327 (1.0)    |  439.7145 (1.0)    |  22.8685 (1.25)   |  435.3772 (1.00)  |   43.5606 (1.31)  |        1;0 | 2.2742 (1.0)   |        5 |          1|
|test_loads[rapidjson_f-100 arrays dict]      |  463.5149 (1.12)   |  504.4733 (1.08)   |  480.9270 (1.09)   |  18.2987 (1.0)    |  478.2495 (1.10)  |   33.2003 (1.0)   |        1;0 | 2.0793 (0.91)  |        5 |          1|
|test_loads[rapidjson_c-100 arrays dict]      |  467.7135 (1.13)   |  525.1420 (1.13)   |  501.0990 (1.14)   |  23.8259 (1.30)   |  511.0303 (1.18)  |   37.5766 (1.13)  |        2;0 | 1.9956 (0.88)  |        5 |          1|
|test_loads[rapidjson_p-100 arrays dict]      |  711.0753 (1.71)   |  766.0270 (1.64)   |  736.3918 (1.67)   |  26.8842 (1.47)   |  730.3733 (1.68)  |   52.4806 (1.58)  |        1;0 | 1.3580 (0.60)  |        5 |          1|

### benchmark '100 arrays dict: serialize': 5 tests

|Name (time in ms)                            |       Min          |       Max          |      Mean          |   StdDev          |    Median         |       IQR          |  Outliers |    OPS          |  Rounds | Iterations|
|---------------------------------------------|--------------------|--------------------|--------------------|-------------------|-------------------|--------------------|-----------|-----------------|---------|-----------|
|test_dumps[rapidjson_nn_f-100 arrays dict]   |  172.8537 (1.0)    |  173.3065 (1.0)    |  173.0847 (1.0)    |   0.1443 (2.12)   |  173.0975 (1.0)   |    0.0469 (1.0)    |       2;2 | 5.7775 (1.0)    |       6 |          1|
|test_dumps[rapidjson_nn_c-100 arrays dict]   |  173.0430 (1.00)   |  173.4205 (1.00)   |  173.2184 (1.00)   |   0.1239 (1.82)   |  173.2011 (1.00)  |    0.0951 (2.03)   |       2;1 | 5.7731 (1.00)   |       6 |          1|
|test_dumps[rapidjson_c-100 arrays dict]      |  235.8061 (1.36)   |  235.9710 (1.36)   |  235.9094 (1.36)   |   0.0679 (1.0)    |  235.9165 (1.36)  |    0.1025 (2.18)   |       1;0 | 4.2389 (0.73)   |       5 |          1|
|test_dumps[rapidjson_p-100 arrays dict]      |  235.8634 (1.36)   |  237.0031 (1.37)   |  236.2245 (1.36)   |   0.4664 (6.87)   |  235.9852 (1.36)  |    0.5457 (11.63)  |       1;0 | 4.2333 (0.73)   |       5 |          1|
|test_dumps[rapidjson_f-100 arrays dict]      |  237.1856 (1.37)   |  265.0171 (1.53)   |  247.7389 (1.43)   |  12.4813 (183.75) |  241.4216 (1.39)  |   20.9180 (445.74) |       1;0 | 4.0365 (0.70)   |       5 |          1|

### benchmark '100 dicts array: deserialize': 5 tests

|Name (time in us)                            |      Min           |      Max           |    Mean          |  StdDev         |    Median         |      IQR         |   Outliers | OPS (Kops/s)        |    Rounds|  Iterations|
|---------------------------------------------|--------------------|--------------------|------------------|-----------------|-------------------|------------------|------------|---------------------|----------|------------|
|test_loads[rapidjson_nn_c-100 dicts array]   |  27.4600 (1.0)     |  82.3340 (1.40)    | 27.8866 (1.0)    |  0.8258 (1.0)   |   27.8060 (1.0)   |   0.1570 (1.0)   |    443;936 |      35.8596 (1.0)  |     24703|           1|
|test_loads[rapidjson_nn_f-100 dicts array]   |  27.9260 (1.02)    | 125.5880 (2.13)    | 28.8076 (1.03)   |  2.9134 (3.53)  |   28.3270 (1.02)  |   0.1990 (1.27)  |   364;1922 |      34.7131 (0.97) |     13856|           1|
|test_loads[rapidjson_f-100 dicts array]      |  32.2330 (1.17)    |  58.8460 (1.0)     | 33.1186 (1.19)   |  2.1873 (2.65)  |   32.6215 (1.17)  |   0.5840 (3.72)  |       9;19 |      30.1945 (0.84) |       342|           1|
|test_loads[rapidjson_c-100 dicts array]      |  32.2790 (1.18)    |  84.0190 (1.43)    | 32.8503 (1.18)   |  2.0177 (2.44)  |   32.6350 (1.17)  |   0.1605 (1.02)  |    311;689 |      30.4411 (0.85) |     13924|           1|
|test_loads[rapidjson_p-100 dicts array]      |  50.6250 (1.84)    | 113.5710 (1.93)    | 51.3652 (1.84)   |  1.4107 (1.71)  |   51.2080 (1.84)  |   0.2500 (1.59)  |    415;682 |      19.4684 (0.54) |     15174|           1|

### benchmark '256 Trues array: deserialize': 5 tests

|Name (time in us)                            |     Min          |       Max          |    Mean          |  StdDev          |  Median          |     IQR          |  Outliers | OPS (Kops/s)          |  Rounds | Iterations|
|---------------------------------------------|------------------|--------------------|------------------|------------------|------------------|------------------|-----------|-----------------------|---------|-----------|
|test_loads[rapidjson_nn_c-256 Trues array]   |  5.4720 (1.0)    |  110.6210 (3.32)   |  5.6453 (1.0)    |  0.4070 (1.00)   |  5.6260 (1.0)    |  0.0480 (1.17)   |  742;4582 |     177.1373 (1.0)    |  116239 |          1|
|test_loads[rapidjson_c-256 Trues array]      |  5.6020 (1.02)   |  101.3130 (3.04)   |  5.7216 (1.01)   |  0.4490 (1.11)   |  5.7010 (1.01)   |  0.0420 (1.02)   |  533;3147 |     174.7758 (0.99)   |  110339 |          1|
|test_loads[rapidjson_f-256 Trues array]      |  5.6310 (1.03)   |   33.3460 (1.0)    |  5.7687 (1.02)   |  0.4057 (1.0)    |  5.7350 (1.02)   |  0.0410 (1.0)    |  721;3547 |     173.3505 (0.98)   |   77888 |          1|
|test_loads[rapidjson_nn_f-256 Trues array]   |  5.8830 (1.08)   |   79.2460 (2.38)   |  6.0703 (1.08)   |  0.6840 (1.69)   |  6.0070 (1.07)   |  0.0510 (1.24)   |  760;4401 |     164.7353 (0.93)   |   75724 |          1|
|test_loads[rapidjson_p-256 Trues array]      |  6.3030 (1.15)   |  121.6370 (3.65)   |  6.4460 (1.14)   |  0.5988 (1.48)   |  6.4050 (1.14)   |  0.0530 (1.29)   |  601;3966 |     155.1357 (0.88)   |   95302 |          1|

### benchmark '256 ascii array: deserialize': 5 tests

|Name (time in us)                            |      Min           |      Max          |     Mean          |  StdDev          |   Median          |     IQR          |  Outliers | OPS (Kops/s)           | Rounds | Iterations|
|---------------------------------------------|--------------------|-------------------|-------------------|------------------|-------------------|------------------|-----------|------------------------|--------|-----------|
|test_loads[rapidjson_nn_c-256 ascii array]   |  33.5800 (1.0)     |  75.3100 (1.0)    |  34.0291 (1.0)    |  0.8575 (1.0)    |  33.9440 (1.0)    |  0.0950 (1.06)   |  455;1122 |      29.3866 (1.0)     |  24085 |          1|
|test_loads[rapidjson_nn_f-256 ascii array]   |  33.8240 (1.01)    | 162.5770 (2.16)   |  34.4227 (1.01)   |  1.8888 (2.20)   |  34.2600 (1.01)   |  0.1500 (1.67)   |   243;923 |      29.0506 (0.99)    |  16691 |          1|
|test_loads[rapidjson_f-256 ascii array]      |  33.8640 (1.01)    |  92.7000 (1.23)   |  34.3083 (1.01)   |  2.1484 (2.51)   |  34.0560 (1.00)   |  0.0900 (1.0)    |   260;814 |      29.1475 (0.99)    |  12805 |          1|
|test_loads[rapidjson_c-256 ascii array]      |  33.9080 (1.01)    | 138.1640 (1.83)   |  35.7549 (1.05)   |  4.9993 (5.83)   |  34.2360 (1.01)   |  0.9640 (10.71)  | 1057;1187 |      27.9682 (0.95)    |  17089 |          1|
|test_loads[rapidjson_p-256 ascii array]      |  51.6020 (1.54)    | 163.7110 (2.17)   |  52.3767 (1.54)   |  3.4894 (4.07)   |  51.9130 (1.53)   |  0.1440 (1.60)   |  291;1555 |      19.0925 (0.65)    |  16900 |          1|

### benchmark '256 doubles array: deserialize': 5 tests

|Name (time in us)                              |      Min          |       Max          |     Mean          |  StdDev          |   Median          |     IQR          |  Outliers | OPS (Kops/s)           | Rounds | Iterations|
|-----------------------------------------------|-------------------|--------------------|-------------------|------------------|-------------------|------------------|-----------|------------------------|--------|-----------|
|test_loads[rapidjson_nn_c-256 doubles array]   |  16.4730 (1.0)    |  154.1860 (1.01)   |  16.9288 (1.0)    |  1.8041 (1.31)   |  16.7620 (1.0)    |  0.1800 (1.15)   |  318;1368 |      59.0710 (1.0)     |  26421 |          1|
|test_loads[rapidjson_nn_f-256 doubles array]   |  16.8560 (1.02)   |  151.9520 (1.0)    |  17.3254 (1.02)   |  1.3765 (1.0)    |  17.2080 (1.03)   |  0.2160 (1.38)   |   250;363 |      57.7188 (0.98)    |  26880 |          1|
|test_loads[rapidjson_c-256 doubles array]      |  64.2160 (3.90)   |  189.7430 (1.25)   |  64.9375 (3.84)   |  4.1525 (3.02)   |  64.6030 (3.85)   |  0.1560 (1.0)    |    53;370 |      15.3994 (0.26)    |   7369 |          1|
|test_loads[rapidjson_f-256 doubles array]      |  64.2260 (3.90)   |  194.2890 (1.28)   |  65.2885 (3.86)   |  4.0041 (2.91)   |  64.8880 (3.87)   |  0.2275 (1.46)   |    87;498 |      15.3166 (0.26)    |   7588 |          1|
|test_loads[rapidjson_p-256 doubles array]      |  65.1430 (3.95)   |  191.4220 (1.26)   |  65.8378 (3.89)   |  2.3332 (1.69)   |  65.6280 (3.92)   |  0.2480 (1.59)   |   280;593 |      15.1888 (0.26)    |  13513 |          1|

### benchmark '256 unicode array: deserialize': 5 tests

|Name (time in us)                              |       Min          |       Max          |      Mean          |   StdDev          |    Median          |     IQR            |Outliers | OPS (Kops/s)            |Rounds | Iterations|
|-----------------------------------------------|--------------------|--------------------|--------------------|-------------------|--------------------|--------------------|---------|-------------------------|-------|-----------|
|test_loads[rapidjson_f-256 unicode array]      |  447.5810 (1.0)    |  825.1410 (1.32)   |  450.3678 (1.0)    |  15.3713 (2.42)   |  448.6985 (1.0)    |  0.8390 (1.24)     |  26;240 |       2.2204 (1.0)      |  1704 |          1|
|test_loads[rapidjson_nn_f-256 unicode array]   |  448.9290 (1.00)   |  788.4770 (1.26)   |  453.5717 (1.01)   |  19.1387 (3.01)   |  449.9220 (1.00)   |  1.5307 (2.27)     |  51;224 |       2.2047 (0.99)     |  1487 |          1|
|test_loads[rapidjson_nn_c-256 unicode array]   |  449.1570 (1.00)   |  768.5950 (1.23)   |  452.4397 (1.00)   |  12.2267 (1.92)   |  450.1970 (1.00)   |  1.3475 (2.00)     |  76;310 |       2.2102 (1.00)     |  2108 |          1|
|test_loads[rapidjson_c-256 unicode array]      |  450.5890 (1.01)   |  888.1580 (1.42)   |  454.0530 (1.01)   |  18.0416 (2.84)   |  451.2685 (1.01)   |  0.6745 (1.0)      |  55;304 |       2.2024 (0.99)     |  1728 |          1|
|test_loads[rapidjson_p-256 unicode array]      |  470.8050 (1.05)   |  624.7800 (1.0)    |  473.4657 (1.05)   |   6.3589 (1.0)    |  472.0950 (1.05)   |  1.5883 (2.35)     |  56;241 |       2.1121 (0.95)     |  2013 |          1|

### benchmark 'apache.json: deserialize': 5 tests

|Name (time in us)                                 |       Min          |         Max            |      Mean          |   StdDev            |    Median          |     IQR            |Outliers        | OPS            |Rounds | Iterations|
|--------------------------------------------------|--------------------|------------------------|--------------------|---------------------|--------------------|--------------------|----------------|----------------|-------|-----------|
|test_loads_sample[rapidjson_c-apache.json]        |  705.6740 (1.0)    |  1,044.7420 (1.16)     |  713.8122 (1.00)   |  23.1400 (2.23)     |  708.5320 (1.0)    |  3.1953 (1.54)     |  50;115  1,400.|9287 (1.00)     |  1083 |          1|
|test_loads_sample[rapidjson_nn_f-apache.json]     |  707.6110 (1.00)   |    901.3720 (1.0)      |  712.3040 (1.0)    |  11.1318 (1.07)     |  710.0840 (1.00)   |  2.2600 (1.09)     |   35;77  1,403.|8949 (1.0)      |  1046 |          1|
|test_loads_sample[rapidjson_nn_c-apache.json]     |  707.8310 (1.00)   |    949.9470 (1.05)     |  712.9454 (1.00)   |  14.9522 (1.44)     |  709.9955 (1.00)   |  2.1280 (1.02)     |  40;110  1,402.|6320 (1.00)     |  1006 |          1|
|test_loads_sample[rapidjson_f-apache.json]        |  711.4380 (1.01)   |    912.5400 (1.01)     |  716.1485 (1.01)   |  10.3901 (1.0)      |  714.0410 (1.01)   |  2.0785 (1.0)      |  40;105  1,396.|3585 (0.99)     |  1089 |          1|
|test_loads_sample[rapidjson_p-apache.json]        |1,096.5980 (1.55)   |  1,415.8250 (1.57)     |1,102.8897 (1.55)   |  18.2262 (1.75)     |1,099.5270 (1.55)   |  4.1995 (2.02)     |   27;33    906.|7090 (0.65)     |   733 |          1|

### benchmark 'canada.json: deserialize': 5 tests

|Name (time in ms)                               |      Min          |      Max          |     Mean          |   StdDev          |   Median          |      IQR            O|utliers     | OPS            |Rounds | Iterations|
|------------------------------------------------|-------------------|-------------------|-------------------|-------------------|-------------------|----------------------|------------|----------------|-------|-----------|
|test_loads_sample[rapidjson_nn_c-canada.json]   |  14.1125 (1.0)    |  40.8366 (1.0)    |  29.2060 (1.00)   |  12.4154 (1.0)    |  38.3138 (1.0)    |  25.1397 (1.0)       |   27;0  34.|2395 (1.00)     |    67 |          1|
|test_loads_sample[rapidjson_nn_f-canada.json]   |  14.2006 (1.01)   |  41.7962 (1.02)   |  29.1343 (1.0)    |  12.8073 (1.03)   |  38.4778 (1.00)   |  25.5885 (1.02)      |   10;0  34.|3238 (1.0)      |    24 |          1|
|test_loads_sample[rapidjson_c-canada.json]      |  43.6217 (3.09)   |  70.4091 (1.72)   |  59.1217 (2.03)   |  12.5862 (1.01)   |  68.0952 (1.78)   |  25.6512 (1.02)      |    9;0  16.|9143 (0.49)     |    23 |          1|
|test_loads_sample[rapidjson_f-canada.json]      |  43.7706 (3.10)   |  70.3776 (1.72)   |  58.4227 (2.01)   |  12.8291 (1.03)   |  67.9360 (1.77)   |  25.1548 (1.00)      |    6;0  17.|1166 (0.50)     |    14 |          1|
|test_loads_sample[rapidjson_p-canada.json]      |  49.0446 (3.48)   |  76.5691 (1.88)   |  64.7160 (2.22)   |  12.9260 (1.04)   |  73.8826 (1.93)   |  25.8240 (1.03)      |    8;0  15.|4521 (0.45)     |    20 |          1|

### benchmark 'complex object: deserialize': 5 tests

|Name (time in us)                            |      Min          |       Max          |      Mean           |  StdDev           |   Median          |     IQR           | Outliers | OPS (Kops/s)            |Rounds | Iterations|
|---------------------------------------------|-------------------|--------------------|---------------------|-------------------|-------------------|-------------------|----------|-------------------------|-------|-----------|
|test_loads[rapidjson_nn_c-complex object]    |  98.7920 (1.0)    |  270.0240 (1.06)   |  100.0089 (1.0)     |  3.2590 (1.0)     |  99.7130 (1.0)    |  0.4290 (1.21)    |  147;332 |       9.9991 (1.0)      |  7050 |          1|
|test_loads[rapidjson_nn_f-complex object]    |  99.7290 (1.01)   |  254.3890 (1.0)    |  101.8566 (1.02)    |  8.1687 (2.51)    | 100.5820 (1.01)   |  0.3555 (1.0)     |  129;493 |       9.8177 (0.98)     |  4632 |          1|
|test_loads[rapidjson_c-complex object]       | 116.1750 (1.18)   |  261.0500 (1.03)   |  118.4774 (1.18)    |  9.3213 (2.86)    | 116.9590 (1.17)   |  0.5580 (1.57)    |  104;422 |       8.4404 (0.84)     |  3846 |          1|
|test_loads[rapidjson_f-complex object]       | 116.6820 (1.18)   |  303.3050 (1.19)   |  118.4544 (1.18)    |  7.8994 (2.42)    | 117.6010 (1.18)   |  0.4242 (1.19)    |   55;213 |       8.4421 (0.84)     |  3621 |          1|
|test_loads[rapidjson_p-complex object]       | 156.1710 (1.58)   |  419.3440 (1.65)   |  158.6774 (1.59)    | 10.9286 (3.35)    | 157.7600 (1.58)   |  0.8410 (2.37)    |   56;289 |       6.3021 (0.63)     |  4886 |          1|

### benchmark 'composite object: deserialize': 5 tests

|Name (time in us)                             |      Min          |       Max          |     Mean          |  StdDev          |   Median          |     IQR          |  Outliers | OPS (Kops/s)           | Rounds | Iterations|
|----------------------------------------------|-------------------|--------------------|-------------------|------------------|-------------------|------------------|-----------|------------------------|--------|-----------|
|test_loads[rapidjson_nn_c-composite object]   |  41.6680 (1.0)    |  186.9480 (1.26)   |  42.5945 (1.0)    |  4.2320 (1.48)   |  42.0150 (1.0)    |  0.1440 (1.07)   |  208;1052 |      23.4772 (1.0)     |  11154 |          1|
|test_loads[rapidjson_nn_f-composite object]   |  42.0070 (1.01)   |  153.8600 (1.04)   |  42.6310 (1.00)   |  2.8664 (1.0)    |  42.3410 (1.01)   |  0.1350 (1.0)    |   196;624 |      23.4571 (1.00)    |  10814 |          1|
|test_loads[rapidjson_f-composite object]      |  49.5120 (1.19)   |  147.7950 (1.0)    |  50.3553 (1.18)   |  3.7382 (1.30)   |  49.9130 (1.19)   |  0.1750 (1.30)   |   131;568 |      19.8589 (0.85)    |   8238 |          1|
|test_loads[rapidjson_c-composite object]      |  50.1780 (1.20)   |  158.8070 (1.07)   |  50.8250 (1.19)   |  3.3515 (1.17)   |  50.5130 (1.20)   |  0.1600 (1.19)   |   107;523 |      19.6754 (0.84)    |   9246 |          1|
|test_loads[rapidjson_p-composite object]      |  66.4950 (1.60)   |  212.1440 (1.44)   |  67.4861 (1.58)   |  3.9212 (1.37)   |  66.9350 (1.59)   |  0.2740 (2.03)   |  216;1299 |      14.8179 (0.63)    |  11797 |          1|

### benchmark 'ctm.json: deserialize': 5 tests

|Name (time in ms)                             |     Min          |      Max          |     Mean          |   StdDev           |  Median          |     IQR            O|utliers     | OPS            |Rounds | Iterations|
|----------------------------------------------|------------------|-------------------|-------------------|--------------------|------------------|---------------------|------------|----------------|-------|-----------|
|test_loads_sample[rapidjson_nn_c-ctm.json]    |  7.8704 (1.0)    |  33.5233 (1.01)   |  13.5606 (1.0)    |  10.2649 (1.03)    |  8.0883 (1.0)    |  0.2753 (1.18)      |  22;24  73.|7429 (1.0)      |    98 |          1|
|test_loads_sample[rapidjson_nn_f-ctm.json]    |  7.9459 (1.01)   |  34.6797 (1.04)   |  14.0420 (1.04)   |  11.0429 (1.11)    |  8.1510 (1.01)   |  0.2339 (1.0)       |    6;6  71.|2150 (0.97)     |    26 |          1|
|test_loads_sample[rapidjson_c-ctm.json]       |  8.6238 (1.10)   |  33.2394 (1.0)    |  14.1749 (1.05)   |   9.9245 (1.0)     |  8.8409 (1.09)   |  0.2555 (1.09)      |  21;22  70.|5474 (0.96)     |    92 |          1|
|test_loads_sample[rapidjson_f-ctm.json]       |  8.6524 (1.10)   |  34.8691 (1.05)   |  14.4448 (1.07)   |  10.4165 (1.05)    |  8.8644 (1.10)   |  0.4084 (1.75)      |  21;21  69.|2291 (0.94)     |    93 |          1|
|test_loads_sample[rapidjson_p-ctm.json]       | 12.8247 (1.63)   |  38.9226 (1.17)   |  18.6672 (1.38)   |  10.5189 (1.06)    | 13.0355 (1.61)   |  0.3047 (1.30)      |  15;16  53.|5699 (0.73)     |    66 |          1|

### benchmark 'github.json: deserialize': 5 tests

|Name (time in us)                               |       Min          |       Max          |      Mean           |  StdDev          |    Median          |     IQR            |Outliers | OPS (Kops/s)            |Rounds | Iterations|
|------------------------------------------------|--------------------|--------------------|---------------------|------------------|--------------------|--------------------|---------|-------------------------|-------|-----------|
|test_loads_sample[rapidjson_nn_c-github.json]   |  337.8400 (1.0)    |  502.6890 (1.21)   |  340.6699 (1.0)     |  6.7109 (1.77)   |  339.6250 (1.0)    |  0.8338 (1.0)      |  77;282 |       2.9354 (1.0)      |  2569 |          1|
|test_loads_sample[rapidjson_nn_f-github.json]   |  338.8910 (1.00)   |  552.9120 (1.33)   |  342.0120 (1.00)    |  8.0093 (2.12)   |  340.7530 (1.00)   |  0.9860 (1.18)     |  80;286 |       2.9239 (1.00)     |  2546 |          1|
|test_loads_sample[rapidjson_c-github.json]      |  346.1330 (1.02)   |  493.7070 (1.18)   |  349.0535 (1.02)    |  5.5681 (1.47)   |  348.0895 (1.02)   |  0.9570 (1.15)     |  67;274 |       2.8649 (0.98)     |  2476 |          1|
|test_loads_sample[rapidjson_f-github.json]      |  347.8150 (1.03)   |  416.6440 (1.0)    |  350.5437 (1.03)    |  3.7818 (1.0)    |  349.8260 (1.03)   |  0.9815 (1.18)     | 111;289 |       2.8527 (0.97)     |  2651 |          1|
|test_loads_sample[rapidjson_p-github.json]      |  466.8610 (1.38)   |  936.0530 (2.25)   |  471.1898 (1.38)    | 18.4494 (4.88)   |  469.3810 (1.38)   |  1.2140 (1.46)     |  19;252 |       2.1223 (0.72)     |  1839 |          1|

### benchmark 'instruments.json: deserialize': 5 tests

|Name (time in ms)                                    |     Min          |      Max          |    Mean          |  StdDev          |  Median          |     IQR            O|utliers      | OPS            |Rounds | Iterations|
|-----------------------------------------------------|------------------|-------------------|------------------|------------------|------------------|---------------------|-------------|----------------|-------|-----------|
|test_loads_sample[rapidjson_nn_c-instruments.json]   |  1.2460 (1.0)    |  27.4478 (14.99)  |  1.3576 (1.03)   |  1.5921 (94.86)  |  1.2514 (1.0)    |  0.0047 (1.0)       |   3;48  736.|6015 (0.97)     |   710 |          1|
|test_loads_sample[rapidjson_nn_f-instruments.json]   |  1.2465 (1.00)   |  23.8052 (13.00)  |  1.3174 (1.0)    |  1.1746 (69.99)  |  1.2516 (1.00)   |  0.0047 (1.01)      |   2;45  759.|0438 (1.0)      |   705 |          1|
|test_loads_sample[rapidjson_f-instruments.json]      |  1.4376 (1.15)   |  22.9988 (12.56)  |  1.5515 (1.18)   |  1.4618 (87.10)  |  1.4463 (1.16)   |  0.0056 (1.21)      |   3;47  644.|5488 (0.85)     |   611 |          1|
|test_loads_sample[rapidjson_c-instruments.json]      |  1.4390 (1.15)   |  25.1383 (13.73)  |  1.5427 (1.17)   |  1.3534 (80.64)  |  1.4564 (1.16)   |  0.0267 (5.74)      |   2;38  648.|2191 (0.85)     |   610 |          1|
|test_loads_sample[rapidjson_p-instruments.json]      |  1.7656 (1.42)   |   1.8312 (1.0)    |  1.7787 (1.35)   |  0.0168 (1.0)    |  1.7730 (1.42)   |  0.0065 (1.40)      |    5;6  562.|1933 (0.74)     |    34 |          1|

### benchmark 'mesh.json: deserialize': 5 tests

|Name (time in ms)                             |     Min          |      Max          |    Mean          |  StdDev          |  Median          |     IQR            O|utliers      | OPS            |Rounds | Iterations|
|----------------------------------------------|------------------|-------------------|------------------|------------------|------------------|---------------------|-------------|----------------|-------|-----------|
|test_loads_sample[rapidjson_nn_f-mesh.json]   |  3.5324 (1.0)    |  28.3195 (1.02)   |  4.5069 (1.0)    |  4.5622 (1.05)   |  3.5567 (1.00)   |  0.1668 (1.03)      |   9;11  221.|8841 (1.0)      |   246 |          1|
|test_loads_sample[rapidjson_nn_c-mesh.json]   |  3.5364 (1.00)   |  27.8498 (1.0)    |  4.5716 (1.01)   |  4.7092 (1.08)   |  3.5500 (1.0)    |  0.1627 (1.0)       |  10;12  218.|7410 (0.99)     |   251 |          1|
|test_loads_sample[rapidjson_c-mesh.json]      |  7.0933 (2.01)   |  32.7126 (1.17)   |  8.1509 (1.81)   |  4.5595 (1.05)   |  7.1754 (2.02)   |  0.1769 (1.09)      |    5;7  122.|6856 (0.55)     |   126 |          1|
|test_loads_sample[rapidjson_f-mesh.json]      |  7.1225 (2.02)   |  28.8747 (1.04)   |  8.1520 (1.81)   |  4.3425 (1.0)    |  7.1895 (2.03)   |  0.1784 (1.10)      |    5;6  122.|6688 (0.55)     |   118 |          1|
|test_loads_sample[rapidjson_p-mesh.json]      |  7.3851 (2.09)   |  32.7411 (1.18)   |  8.5458 (1.90)   |  4.9490 (1.14)   |  7.4754 (2.11)   |  0.1740 (1.07)      |    5;8  117.|0159 (0.53)     |   123 |          1|

### benchmark 'truenull.json: deserialize': 5 tests

|Name (time in us)                                 |      Min          |       Max          |     Mean          |  StdDev          |   Median          |     IQR           | Outliers | OPS (Kops/s)           | Rounds | Iterations|
|--------------------------------------------------|-------------------|--------------------|-------------------|------------------|-------------------|-------------------|----------|------------------------|--------|-----------|
|test_loads_sample[rapidjson_nn_c-truenull.json]   |  49.4600 (1.0)    |  109.2150 (1.13)   |  50.0595 (1.0)    |  1.6238 (1.17)   |  49.8600 (1.0)    |  0.2240 (2.31)    |  311;624 |      19.9762 (1.0)     |  13632 |          1|
|test_loads_sample[rapidjson_f-truenull.json]      |  49.8630 (1.01)   |   96.7070 (1.0)    |  50.5642 (1.01)   |  1.3826 (1.0)    |  50.4340 (1.01)   |  0.1110 (1.14)    | 226;1628 |      19.7769 (0.99)    |  11119 |          1|
|test_loads_sample[rapidjson_c-truenull.json]      |  49.9230 (1.01)   |  115.8070 (1.20)   |  50.4689 (1.01)   |  1.5109 (1.09)   |  50.3130 (1.01)   |  0.2570 (2.65)    |  193;407 |      19.8142 (0.99)    |  10702 |          1|
|test_loads_sample[rapidjson_nn_f-truenull.json]   |  50.4790 (1.02)   |  112.1650 (1.16)   |  50.9551 (1.02)   |  1.5255 (1.10)   |  50.7620 (1.02)   |  0.0970 (1.0)     | 340;1847 |      19.6251 (0.98)    |  13943 |          1|
|test_loads_sample[rapidjson_p-truenull.json]      |  50.8280 (1.03)   |  135.7310 (1.40)   |  51.3831 (1.03)   |  1.6023 (1.16)   |  51.2180 (1.03)   |  0.1220 (1.26)    | 264;1094 |      19.4616 (0.97)    |   9435 |          1|

### benchmark 'tweet.json: deserialize': 5 tests

|Name (time in us)                              |      Min           |      Max          |     Mean          |  StdDev          |   Median          |     IQR           | Outliers | OPS (Kops/s)           | Rounds | Iterations|
|-----------------------------------------------|--------------------|-------------------|-------------------|------------------|-------------------|-------------------|----------|------------------------|--------|-----------|
|test_loads_sample[rapidjson_nn_c-tweet.json]   |  28.5530 (1.0)     |  87.3270 (1.12)   |  29.0590 (1.0)    |  1.0791 (1.0)    |  28.9760 (1.0)    |  0.1510 (1.03)    |  335;877 |      34.4127 (1.0)     |  22550 |          1|
|test_loads_sample[rapidjson_nn_f-tweet.json]   |  29.0080 (1.02)    |  82.9770 (1.06)   |  29.4918 (1.01)   |  1.4135 (1.31)   |  29.3630 (1.01)   |  0.1470 (1.0)     |  293;754 |      33.9078 (0.99)    |  16418 |          1|
|test_loads_sample[rapidjson_f-tweet.json]      |  29.6430 (1.04)    |  78.0520 (1.0)    |  30.2539 (1.04)   |  1.1218 (1.04)   |  30.1300 (1.04)   |  0.1623 (1.10)    |  266;911 |      33.0536 (0.96)    |  14525 |          1|
|test_loads_sample[rapidjson_c-tweet.json]      |  29.7990 (1.04)    |  81.5400 (1.04)   |  30.3492 (1.04)   |  1.1033 (1.02)   |  30.2340 (1.04)   |  0.1520 (1.03)    |  299;946 |      32.9498 (0.96)    |  17408 |          1|
|test_loads_sample[rapidjson_p-tweet.json]      |  38.3150 (1.34)    | 148.8100 (1.91)   |  38.9714 (1.34)   |  1.7235 (1.60)   |  38.8290 (1.34)   |  0.2030 (1.38)    |  186;626 |      25.6599 (0.75)    |  12801 |          1|

### benchmark 'twitter.json: deserialize': 5 tests

|Name (time in ms)                                |     Min          |      Max          |    Mean          |  StdDev          |  Median          |     IQR            O|utliers      | OPS            |Rounds | Iterations|
|-------------------------------------------------|------------------|-------------------|------------------|------------------|------------------|---------------------|-------------|----------------|-------|-----------|
|test_loads_sample[rapidjson_nn_c-twitter.json]   |  4.0228 (1.0)    |  28.6704 (1.06)   |  4.5468 (1.0)    |  3.3973 (1.06)   |  4.0389 (1.0)    |  0.0199 (1.0)       |   4;39  219.|9340 (1.0)      |   199 |          1|
|test_loads_sample[rapidjson_nn_f-twitter.json]   |  4.0264 (1.00)   |  28.0712 (1.04)   |  4.5509 (1.00)   |  3.3484 (1.05)   |  4.0441 (1.00)   |  0.0376 (1.89)      |   4;25  219.|7364 (1.00)     |   193 |          1|
|test_loads_sample[rapidjson_f-twitter.json]      |  4.1626 (1.03)   |  27.0811 (1.0)    |  4.6711 (1.03)   |  3.1971 (1.0)    |  4.1845 (1.04)   |  0.0283 (1.42)      |   4;29  214.|0809 (0.97)     |   190 |          1|
|test_loads_sample[rapidjson_c-twitter.json]      |  4.1643 (1.04)   |  28.7822 (1.06)   |  4.7883 (1.05)   |  3.6020 (1.13)   |  4.1796 (1.03)   |  0.0556 (2.79)      |   5;11  208.|8408 (0.95)     |   196 |          1|
|test_loads_sample[rapidjson_p-twitter.json]      |  5.1758 (1.29)   |  29.7481 (1.10)   |  5.8116 (1.28)   |  3.7307 (1.17)   |  5.1955 (1.29)   |  0.0488 (2.45)      |   4;11  172.|0700 (0.78)     |   163 |          1|

lelit commented:
As said, I will try some idea to see if I can eliminate that overhead with some trick at the `C++` level.

GoodWasHere commented:
Yeah, the overhead is epic.
But that's not why we love Python )

Spitfire1900 commented:
It's not an empty method if it does not exist. Linters and language servers to not assume the existence of `__div__` on function unless the class implements it. Also none of these methods are intended to be called by the end user, they're intended to be called by the inherited `__call__()`'s cpp implementation.


2bf122ef8f780a853646705f06ce58124434979f is a more suitable commit to merge than the current HEAD, 9a2580ba8b5e75b19b6498c561407a34f3212e77.

lelit commented:
FYI, you can find updated benchmarks [here](https://python-rapidjson.readthedocs.io/en/latest/benchmarks.html#tables).

Please drop me a note when you feel this is ready to be merged.

GoodWasHere commented:
@lelit @Spitfire1900 
Since `Encoder` and `Decoder` will still not have the above methods, the only thing left to decide is whether `.pyi` should contain their signature.

What is your opinion?

Spitfire1900 commented:
It should not include their signature.

This suggests to the end user that the base implementation contains these methods. They only exist if a sub-type of Encoder/Decoder implements them and a good LSP will "see" the sub-types implementation (see https://github.com/python-rapidjson/python-rapidjson/pull/204#issuecomment-2171932637).

Additionally, these methods act like dunder methods in practice; they are not intended for the end user to call, only the inherited `__call__()` method is intended to call them.

Spitfire1900 commented:
I have an adjacent hot take to this issue, the optional methods are not intended for the end user to cal they should be implemented as _dunder_ methods instead of the current model. E.g. `Decoder.end_array()` should instead be `Decoder.__rapidjson_end_array__()`.

lelit commented:
Yes, those are **optional** methods, for which there are no implementation in the base classes, and the current code is the simplest and efficient way to support them.

OTOH, I don't really get the point made about their name. There are many cases of such *optional* methods even in the standard library, and while the majority are _dunder_-named, some of them are not, despite not being prominently an user-callable-thing. What would be the advantage of renaming them as proposed?

Spitfire1900 commented:
Thinking on it, the rename isn't worth it vs not breaking the existing API, despite my opinions.

As mentioned, this PR's 2bf122e commit is suitable to merge, but not HEAD.

lelit commented:
Ok, I merged this (omitting the `stubtest_allowlist.txt added` commit), with minor modifications for consistency with the existing codebase.
Thanks a lot to both of you!

lelit commented:
Sigh, it seems I introduced an error with commit 47bf0eedf23ab0cade365781192a962e033e00ba: trying `stubtest` on a fresh install, it said that a the `__rapidjson_exact_version__ was missing, so I added it...
I wonder why it complains in the CI test...

lelit commented:
Ok, found out the issue, fixed in 3d84f0a7453d57e872cd695643bec9b8a3f89659.

Spitfire1900 commented:
@lelit thanks a lot! Do you think this could get released on PyPi soon?

lelit commented:
Yes, sure.
